crest/Assets/Crest/Crest/Scripts/Helpers/UnderwaterPostProcess.cs

```diff
@@ -0,0 +1,244 @@
+            }
+
+            bool definitelyAboveTheWater = false;
+            {
+                float oceanHeight = OceanRenderer.Instance.transform.position.y;
+                float maxOceanVerticalDisplacement = OceanRenderer.Instance.MaxVertDisplacement * 0.5f;
+                float cameraHeight = _mainCamera.transform.position.y;
+                definitelyAboveTheWater = (cameraHeight - maxOceanVerticalDisplacement) >= oceanHeight;
+            }
+
+            if (GL.wireframe || definitelyAboveTheWater)
+            {
+                Graphics.Blit(source, target);
+                _oceanChunksToRender.Clear();
+                return;
+            }
+
+            if (_textureMask == null)
+            {
+                _textureMask = new RenderTexture(source);
+                _textureMask.name = "Ocean Mask";
+                // @Memory: We could investigate making this an 8-bit texture
+                // instead to reduce GPU memory usage.
+                _textureMask.format = RenderTextureFormat.RHalf;
+                _textureMask.Create();
+
+                _depthBuffer = new RenderTexture(source);
+                _depthBuffer.name = "Ocean Mask Depth";
+                _depthBuffer.format = RenderTextureFormat.Depth;
+                _depthBuffer.Create();
+            }
+
+            // Get all ocean chunks and render them using cmd buffer, but with mask shader
+            _commandBuffer.SetRenderTarget(_textureMask.colorBuffer, _depthBuffer.depthBuffer);
+            _commandBuffer.ClearRenderTarget(true, true, Color.white * UNDERWATER_MASK_NO_MASK);
+            _commandBuffer.SetViewProjectionMatrices(_mainCamera.worldToCameraMatrix, _mainCamera.projectionMatrix);
+
+            foreach (var chunk in _oceanChunksToRender)
+            {
+                _commandBuffer.DrawRenderer(chunk, _oceanMaskMaterial);
+            }
+
+            {
+                // The same camera has to perform post-processing when performing
+                // VR multipass rendering so don't clear chunks until we have
+                // renderered the right eye.
+                // This was the approach recommended by Unity's post-processing
+                // lead Thomas Hourdel.
+                bool saveChunksToRender =
+                    XRSettings.enabled &&
+                    XRSettings.stereoRenderingMode == XRSettings.StereoRenderingMode.MultiPass &&
+                    _mainCamera.stereoActiveEye != Camera.MonoOrStereoscopicEye.Right;
+
+                if (!saveChunksToRender) _oceanChunksToRender.Clear();
+            }
+
+            _underwaterPostProcessMaterial.CopyPropertiesFromMaterial(OceanRenderer.Instance.OceanMaterial);
+
+            if (_viewOceanMask)
+            {
+                _underwaterPostProcessMaterial.EnableKeyword(DEBUG_VIEW_OCEAN_MASK);
+            }
+            else
+            {
+                _underwaterPostProcessMaterial.DisableKeyword(DEBUG_VIEW_OCEAN_MASK);
```

Comment:
I'm not sure what the overhead of calling DisableKeyword each frame might be. i guess it might cache it internally and early out cheaply?

Reply:
I would really hope so. We could profile this and check?

Reply:
Yes that would be good. I did a very quick pass with the profiler and there's a fair amount of time being spent in this function - around 0.3ms-ish on average, but a number of times 0.5ms, highest 0.7ms).  i'll add a todo to poke around disabling chunks of the function to determine where this time is going..

Reply:
I didnt notice a huge overhead calling this. (the lions share of hte cost seemed to be in submitting the ocean tiles)